File: LeetCode/HashMap/JavaScript/groupAnagrams.js
// Given an array of strings strs, group the
// anagrams
//  together. You can return the answer in any order.

// Example 1:

// Input: strs = ["eat","tea","tan","ate","nat","bat"]

// Output: [["bat"],["nat","tan"],["ate","eat","tea"]]

// Explanation:

// There is no string in strs that can be rearranged to form "bat".
// The strings "nat" and "tan" are anagrams as they can be rearranged to form each other.
// The strings "ate", "eat", and "tea" are anagrams as they can be rearranged to form each other.
// Example 2:

// Input: strs = [""]

// Output: [[""]]

// Example 3:

// Input: strs = ["a"]

// Output: [["a"]]

// Constraints:

// 1 <= strs.length <= 104
// 0 <= strs[i].length <= 100
// strs[i] consists of lowercase English letters.

// TC: O(n*m), we're visiting each strings in the strs array and for each string, we're visiting each chars
// SC: O(n*m), we're creating a map of word count for each string and for each of the word count, we append all the strings that match the word count

/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function (strs) {
  let select = 2;

  if (select === 1) {
    console.log("Using groupAnagrams1()");
    return groupAnagrams1(strs);
  } else {
    console.log("Using groupAnagrams2()");
    return groupAnagrams2(strs);
  }
};

var groupAnagrams1 = function (strs) {
  let res = {};

  for (let s of strs) {
    let count = new Array(26).fill(0); // we need an array of size 26 to store the counts for each chars
    // console.log(s)
    for (let c of s) {
      // Let's count the chars in each strings
      // What were doing here is to map a char to a certain value
      // for example, "d" = 100, "a" = 97 => d-a = 100 - 97 = 3 => "d"
      // here "d" is mapped to 3, and we count the occurence of d, which is mapped to index 3.
      // console.log(c)
      count[c.charCodeAt(0) - "a".charCodeAt(0)] += 1;
    }

    // console.log(count.join())
    // We need to convert count to a string so we can search our hashmap
    // The whole count array will become our key in the hashmap.
    const countStr = count.join();

    // Check first if we have the count string in our hashmap
    // if not, create it and for the value, create an array and push the string s
    // If so, get the k/v pair and push s to the values array
    if (!(countStr in res)) {
      res[countStr] = [];
    }

    res[countStr].push(s);
  }
  // console.log(Object.values(res))
  return Object.values(res);
};

var groupAnagrams2 = function (strs) {
  let res = {};

  for (let s of strs) {
    //  Convert the string to a char array and sort them.
    let charsArr = s.split("");
    charsArr.sort();

    // We need to convert count to a string so we can search our hashmap
    // The whole count array will become our key in the hashmap.
    const countStr = charsArr.join("");

    // Check first if we have the count string in our hashm
    // if not, create it and for the value, create an array and push the string s
    // If so, get the k/v pair and push s to the values array
    if (!(countStr in res)) {
      res[countStr] = [];
    }

    res[countStr].push(s);
  }
  // console.log(Object.values(res))
  return Object.values(res);
};

var testSolution = (input) => {
  console.log("Input: strs: " + JSON.stringify(input["strs"]));
  console.log("Expected: " + JSON.stringify(input["expected"]));
  let res = groupAnagrams(input["strs"]);
  console.log("Result: " + JSON.stringify(res));
  console.log(isEqual(res, input["expected"]) ? "PASS" : "FAIL");
  console.log("-".repeat(50));
};

var isEqual = (arr1, arr2) => {
  if (arr1.length !== arr2.length) {
    return false;
  }

  // Sort inner arrays
  const sortedArr1 = arr1.map((innerArr) => [...innerArr].sort());
  const sortedArr2 = arr2.map((innerArr) => [...innerArr].sort());

  // Sort outer arrays based on stringified inner arrays for consistent comparison
  sortedArr1.sort((a, b) => JSON.stringify(a).localeCompare(JSON.stringify(b)));
  sortedArr2.sort((a, b) => JSON.stringify(a).localeCompare(JSON.stringify(b)));

  // Deep compare the sorted matrices
  for (let i = 0; i < sortedArr1.length; i++) {
    if (sortedArr1[i].length !== sortedArr2[i].length) {
      return false;
    }
    for (let j = 0; j < sortedArr1[i].length; j++) {
      if (sortedArr1[i][j] !== sortedArr2[i][j]) {
        return false;
      }
    }
  }

  return true;
};

// =======================================
// Test Vectors
// =======================================

let input = {
  strs: ["eat", "tea", "tan", "ate", "nat", "bat"],
  expected: [["bat"], ["nat", "tan"], ["ate", "eat", "tea"]],
};

testSolution(input);

input = {
  strs: [""],
  expected: [[""]],
};

testSolution(input);

input = {
  strs: ["a"],
  expected: [["a"]],
};

testSolution(input);

File: LeetCode/HashMap/java/groupAnagrams.java
package LeetCode.HashMap.java;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

record groupAnagramsRecord(String[] strs, List<List<String>> expected) {
}

public class groupAnagrams {
    public List<List<String>> groupAnagramsSolution(String[] strs) {
        int select = 2;

        if (select == 1) {
            System.out.println("Using groupAnagramsSolution1()");
            return groupAnagramsSolution1(strs);
        } else {
            System.out.println("Using groupAnagramsSolution2()");
            return groupAnagramsSolution2(strs);
        }
    }

    public List<List<String>> groupAnagramsSolution1(String[] strs) {
        Map<String, List<String>> res = new HashMap<>();

        for (String s : strs) {
            // we need an array of size 26 to store the counts for each chars
            int[] count = new int[26];
            Arrays.fill(count, 0);

            // Let's count the chars in each strings
            for (char c : s.toCharArray()) {
                // What were doing here is to map a char to a certain value
                // for example, "d" = 100, "a" = 97 => d-a = 100 - 97 = 3 => "d"
                // here "d" is mapped to 3, and we count the occurence of d, which is mapped to
                // index 3.
                char idx = (char) (c - "a".charAt(0));
                count[idx] += 1;
            }

            // We need to convert count array to a string so we can search our hashmap
            // The whole count array will become our key in the hashmap.
            String countStr = Arrays.toString(count);

            // Check first if we have the count string in our hashmap
            // if not, create it and for the value, create an array and push the string s
            // If so, get the k/v pair and push s to the values array
            List<String> tmp = res.getOrDefault(countStr, new ArrayList<>());
            tmp.add(s);
            res.put(countStr, tmp);
        }

        return res.values().stream().toList();
    }

    public List<List<String>> groupAnagramsSolution2(String[] strs) {
        // What we need to do here is go through each string in strs.
        // Group each strings if they have the same number of letters and the same
        // letters.
        Map<String, List<String>> res = new HashMap<>();

        for (String s : strs) {
            // Convert the string to a char array and sort them.
            char[] chars = s.toCharArray();
            Arrays.sort(chars);

            // We want to use this sorted char array as key to our res map.
            // We need to convet it to string.
            String sortedChars = String.valueOf(chars);

            // Check if our sorted chars string key exists in our res map.
            // If not, initialize it with an empty arraylist
            // If so, get the existing arraylist and append the string s.
            List<String> tmp = res.getOrDefault(sortedChars, new ArrayList<>());
            tmp.add(s);
            res.put(sortedChars, tmp);
        }

        return res.values().stream().toList();
    }

    public static void main(String[] args) {
        groupAnagramsRecord input = new groupAnagramsRecord(
                new String[] { "eat", "tea", "tan", "ate", "nat", "bat" },
                List.of(List.of("bat"),
                        List.of("tan", "nat"),
                        List.of("eat", "tea", "ate")));

        groupAnagrams.testSolution(input);

        input = new groupAnagramsRecord(new String[] { "" }, List.of(List.of("")));
        groupAnagrams.testSolution(input);

        input = new groupAnagramsRecord(new String[] { "a" }, List.of(List.of("a")));
        groupAnagrams.testSolution(input);

    }

    private static void testSolution(groupAnagramsRecord input) {

        System.out.println("Input: strs: " + Arrays.toString(input.strs()));
        System.out.println("Expected: " + input.expected().toString());
        List<List<String>> res = new groupAnagrams().groupAnagramsSolution(input.strs());
        System.out.println("Result: " + res.toString());
        System.out.println(groupAnagrams.isEqual(res, input.expected()) ? "PASS" : "FAIL");
        System.out.println("-".repeat(50));
    }

    private static boolean isEqual(List<List<String>> l1, List<List<String>> l2) {
        if (l1.size() != l2.size()) {
            return false;
        }

        // We'll use set to compare the two List<List<String>> input
        // We'll have a Set<Set<String>>

        Set<Set<String>> canonicalList1 = new HashSet<>();

        // Here for each innerlist of the input, add it as a new hashset and append to
        // the outer hashset
        for (List<String> innerList : l1) {
            canonicalList1.add(new HashSet<>(innerList));
        }

        // Here for each innerlist of the input, add it as a new hashset and append to
        // the outer hashset
        Set<Set<String>> canonicalList2 = new HashSet<>();
        for (List<String> innerList : l2) {
            canonicalList2.add(new HashSet<>(innerList));
        }

        // We then compare the outer hashsets
        return canonicalList1.equals(canonicalList2);
    }
}
